### PR TITLE
[Zen2] Best-effort cluster formation if unconfigured

### PR DESCRIPTION
**Pull request by DaveCTurner**
In real deployments it is important that clusters are properly configured to
avoid accidentally forming multiple independent clusters at cluster
bootstrapping time. However we also expect to be able to unpack Elasticsearch
and start up one or more nodes without any up-front configuration, and have
them do their best to find each other and form a cluster after a few seconds.

This change adds a delayed automatic bootstrapping process to nodes that start
up with no relevant settings set to support the desired out-of-the-box
experience without compromising safety in properly-configured deployments.

### CODE REVIEW

**Review by ywelsch — COMMENTED** [review]
A few minor comments, looks good o.w.

File: server/src/main/java/org/elasticsearch/bootstrap/BootstrapChecks.java [hunk 1]
```diff
@@ -713,4 +715,13 @@ boolean isAllPermissionGranted() {
 
     }
 
+    static class DiscoveryConfiguredCheck implements BootstrapCheck {
+        @Override
+        public BootstrapCheckResult check(BootstrapContext context) {
+            if (ClusterBootstrapService.discoveryIsConfigured(context.settings)) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/bootstrap/BootstrapChecks.java]: This should only apply if `discovery.type` is `zen2` - particularly, `single-node` discovery shouldn't yield this warning.

Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/bootstrap/BootstrapChecks.java]: As it stands, we do not have a `discovery.type` of `zen2` :) I will wait for that before proceeding.

File: server/src/main/java/org/elasticsearch/bootstrap/BootstrapChecks.java [hunk 2]
```diff
@@ -713,4 +715,13 @@ boolean isAllPermissionGranted() {
...
+                return BootstrapCheckResult.success();
+            }
+            return BootstrapCheckResult.failure("the default discovery settings are unsuitable for production use");
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/bootstrap/BootstrapChecks.java]: Explicitly mention the settings which need to be set?

File: server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java [hunk 3]
```diff
@@ -57,22 +62,82 @@
...
         assert running == false;
         running = true;
 
-        if (initialMasterNodeCount > 0 && transportService.getLocalNode().isMasterNode()) {
+        if (transportService.getLocalNode().isMasterNode() == false) {
+            return;
+        }
+
+        if (unconfiguredBootstrapTimeout != null) {
+            logger.info("no discovery configuration found, will perform best-effort cluster bootstrapping after [{}] " +
```
Comment (ywelsch) [inline server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java]: can we avoid this message if the cluster is already bootstrapped?

**Review by ywelsch — CHANGES_REQUESTED** [review]

File: server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java (hunk 3 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java]: We can avoid the whole cluster bootstrap service in that case. I pushed 43fdbd00f5f2e737c215c562af1ab4bd38ca1c51.

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by DaveCTurner** [comment]
@elasticmachine please run the Gradle build tests 2

**Pull request merged**